PetAdoptions/pethistory/index.js

```diff
@@ -0,0 +1,237 @@
+    let adoptionData;
+    try {
+        adoptionData = JSON.parse(record.body);
+        console.log('Parsed adoption data:', adoptionData);
+    } catch (error) {
+        console.error('Invalid JSON in message:', messageId, error);
+        throw new Error(`Invalid JSON in message ${messageId}: ${error.message}`);
+    }
+    // Validate message schema
+    const { error, value } = adoptionMessageSchema.validate(adoptionData);
+    if (error) {
+        console.error('Schema validation failed for message:', messageId, error.details);
+        throw new Error(`Schema validation failed for message ${messageId}: ${error.message}`);
+    }
+    const validatedData = value;
+    try {
+        console.log(`Processing adoption - Transaction: ${validatedData.transactionId}, Pet: ${validatedData.petId}, User: ${validatedData.userId}`);
+        // Write adoption history to database
+        await writeAdoptionHistoryToDatabase(validatedData);
+
+        console.log(`Successfully processed adoption for pet ${validatedData.petId} by user ${validatedData.userId}`);
+
+        return {
+            messageId: messageId,
+            transactionId: validatedData.transactionId,
+            status: 'success'
+        };
+
+    } catch (error) {
+        console.error(`Failed to process adoption for pet ${validatedData.petId}:`, error);
+        throw error;
+    }
+}
+
+/**
+ * Write adoption history record to PostgreSQL database
+ */
+async function writeAdoptionHistoryToDatabase(adoptionData) {
+    let client;
+
+    try {
+        console.log(`Writing adoption history to database: ${adoptionData.transactionId}`);
+
+        // Get database connection details from Secrets Manager
+        const dbConfig = await getDatabaseConfig();
+
+        // Create database connection
+        client = new Client({
+            host: dbConfig.host,
+            port: dbConfig.port,
+            database: dbConfig.dbname,
+            user: dbConfig.username,
+            password: dbConfig.password,
+            ssl: false, // Adjust based on your RDS configuration
+            connectionTimeoutMillis: 5000,
+            query_timeout: 10000
+        });
+
+        await client.connect();
+        console.log('Connected to database successfully');
+
+        // Optimistic approach: try to insert first
+        const insertQuery = `
+            INSERT INTO transaction_history (pet_id, transaction_id, adoption_date, user_id, created_at)
+            VALUES ($1, $2, $3, $4, $5)
+        `;
+
+        const values = [
+            adoptionData.petId,
+            adoptionData.transactionId,
+            new Date(adoptionData.adoptionDate),
+            adoptionData.userId,
+            new Date() // Current timestamp for history record
+        ];
+
+        const result = await client.query(insertQuery, values);
+        console.log(`Successfully inserted adoption history record: ${adoptionData.transactionId}, rows affected: ${result.rowCount}`);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Log Injection occurs when untrusted user input is directly written to log files without proper sanitization. This can allow attackers to manipulate log entries, potentially leading to security issues like log forging or cross-site scripting. To prevent this, always sanitize user input using `encodeURIComponent()` or `DOMPurify.sanitize()` before logging.  Learn more - https://cwe.mitre.org/data/definitions/117.html